### PR TITLE
Add algorithmic optimizer to convert Log(Softmax(x)) to LogSoftmax(x)

### PR DESCRIPTION
**Pull request by lgeiger**
This PR adds an algorithmic optimizer which converts `Log(Softmax(x))` to `LogSoftmax(x)`. [`LogSoftmax`](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/log-softmax) is numerically more stable and may be a bit faster in some cases.

This could be expanded in the future to also optimize `log(softmax(x) * y) = logsoftmax(x) + log(y)` and `log(softmax(x) / y) = logsoftmax(x) - log(y)`.

### CODE REVIEW

**Comment by hgadig** [comment]
@rmlarsen  Could you PTAL and approve or suggest changes(if required).

**Comment by lgeiger** [comment]
Rebased since #25300 introduced some merge conflicts.

**Comment by lgeiger** [comment]
@hgadig Any updates on this?

**Comment by hgadig** [comment]
> @hgadig Any updates on this?

Waiting for the review. Also, this is assigned to rthadur and he will be taking care of this PR(fyi). Thanks !

**Review by ezhulenev — APPROVED** [review]

**Comment by ezhulenev** [comment]
Thanks! Sorry for long review cycle.

**Comment by lgeiger** [comment]
> Thanks! Sorry for long review cycle.

Thanks @ezhulenev, you've been really helpful with getting my fixes merged.

**Pull request merged**
